### PR TITLE
RavenDB-10891 make the compaction of a database more robust

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: src/Raven.Server/Utils/IOExtensions.cs [hunk 1]
```diff
@@ -26,6 +26,40 @@ public static void DeleteFile(string file)
...
+            }
+            catch
+            {
+                return false;
+            }
+            finally
+            {
+                if (tmpFileName != null)
+                {
+                    File.Delete(tmpFileName);
```
Comment (ayende) [inline src/Raven.Server/Utils/IOExtensions.cs:58]: Need error handling here, you might fail to write the file, then fail again when trying to delete it.

Comment (talweiss1982) [reply src/Raven.Server/Utils/IOExtensions.cs:58]: @ayende okay, I was wondering if I should also delete the folder if it was created by the exam code? It is likely that we will write to that folder so it is redundant, on one hand but it is an unexpected side effect on the other hand.

Comment (ayende) [reply src/Raven.Server/Utils/IOExtensions.cs:58]: No, don't delete the directory

Comment (talweiss1982) [reply src/Raven.Server/Utils/IOExtensions.cs:58]: arrg already send the PR, wait a sec.

File: src/Raven.Server/Documents/CompactDatabaseTask.cs [hunk 2]
```diff
@@ -87,28 +104,39 @@ public CompactDatabaseTask(ServerStore serverStore, string database, Cancellatio
...
-                    var oldConfigPath = new PathSetting(basePath + "-old").Combine("Configuration");
-                    var newConfigPath = new PathSetting(basePath).Combine("Configuration");
-                    IOExtensions.MoveDirectory(oldConfigPath.FullPath, newConfigPath.FullPath);
-                }
-                catch (Exception e)
-                {
-                    throw new InvalidOperationException($"Failed to execute compaction for {_database}", e);
-                }
-                finally
+                    SwitchDatabaseDirectories(basePath, tmpDirectory, compactDirectory);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/CompactDatabaseTask.cs:108]: I wonder if we should call `EnsureReadWritePermissionForDirectory` again here before `SwitchDatabaseDirectories`? The compaction can take hours for huge dbs, so just to ensure nothing has changed meanwhile and we still have an access

Comment (ayende) [reply src/Raven.Server/Documents/CompactDatabaseTask.cs:108]: Might be a good idea, but if they changed the permissions midway, that is PEBKAC

**Pull request merged**